crates/starknet-devnet/tests/test_simulate_transactions.rs

```diff
@@ -479,4 +493,568 @@ mod simulation_tests {
+        }
+    }
+
+    #[tokio::test]
+    async fn simulate_with_max_fee_exceeding_account_balance_returns_error_if_fee_charge_is_not_skipped()
+     {
+        let devnet = BackgroundDevnet::spawn().await.expect("Could not start Devnet");
+        let (sierra_artifact, casm_hash) =
+            get_flattened_sierra_contract_and_casm_hash(CAIRO_1_PANICKING_CONTRACT_SIERRA_PATH);
+
+        let (signer, account_address) = devnet.get_first_predeployed_account().await;
+
+        let mut account = SingleOwnerAccount::new(
+            &devnet.json_rpc_client,
+            signer,
+            account_address,
+            constants::CHAIN_ID,
+            ExecutionEncoding::New,
+        );
+        account.set_block_id(BlockId::Tag(BlockTag::Latest));
+
+        let declaration = account
+            .declare_v3(Arc::new(sierra_artifact), casm_hash)
+            .gas(u64::MAX)
+            .gas_price(u128::MAX);
+
+        match declaration.simulate(false, false).await.unwrap_err() {
+            AccountError::Provider(ProviderError::StarknetError(
+                StarknetError::TransactionExecutionError(TransactionExecutionErrorData {
+                    execution_error,
+                    ..
+                }),
+            )) => {
+                assert_contains(
+                    &execution_error,
+                    "Account balance is not enough to cover the transaction cost.",
+                );
+            }
+            other => panic!("Unexpected error {other:?}"),
+        }
+
+        // should not fail because fee transfer is skipped
+        declaration.simulate(false, true).await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn simulate_v3_deploy_account_declare_deploy_via_invoke_to_udc_happy_path() {
+        let devnet = BackgroundDevnet::spawn_with_additional_args(&["--account-class", "cairo1"])
+            .await
+            .expect("Could not start Devnet");
+
+        let new_account_private_key = Felt::from(7777);
+        let signer =
+            LocalWallet::from_signing_key(SigningKey::from_secret_scalar(new_account_private_key));
+
+        let latest = BlockId::Tag(BlockTag::Latest);
+
+        let public_key = signer.get_public_key().await.unwrap();
+        let salt = Felt::from_hex_unchecked("0x123");
+        let account_class_hash = Felt::from_hex_unchecked(CAIRO_1_ACCOUNT_CONTRACT_SIERRA_HASH);
+
+        let resource_bounds = ResourceBoundsMapping {
+            l1_gas: ResourceBounds { max_amount: 0, max_price_per_unit: 0 },
+            l2_gas: ResourceBounds { max_amount: 0, max_price_per_unit: 0 },
+        };
+
+        let nonce_data_availability_mode = DataAvailabilityMode::L1;
+        let fee_data_availability_mode = DataAvailabilityMode::L1;
+        let is_query = true;
+        let chain_id = devnet.json_rpc_client.chain_id().await.unwrap();
+        let paymaster_data = vec![];
+        let tip = 0;
+        let gas = 0;
```

Comment:
Why does it work with gas set to 0 and validation not skipped?

Reply:
https://github.com/0xSpaceShard/starknet-devnet-rs/commit/df3537e8597cd5add216747716dbfeed1ff9a2fb